File: alatting_website/logic/poster_render.py
# coding=utf-8

import re
from decimal import Decimal
from django.template.loader import render_to_string


class PosterRender:
    @classmethod
    def json_to_html(cls, root, context):
        pages = root['pages']
        page = pages[0]
        html = ''
        for element in page:
            typ = element['type']
            if typ == 'background':
                html += cls.render_background(element, context) + '\n'
            elif typ == 'text':
                html += cls.render_text(element, context) + '\n'
            elif typ == 'button':
                html += cls.render_button(element, context) + '\n'
            elif typ == 'image':
                html += cls.render_image(element, context) + '\n'
            elif typ == 'slider':
                html += cls.render_slider(element, context) + '\n'
            elif typ == 'music':
                html += cls.render_music(element, context) + '\n'
            elif typ == 'video':
                html += cls.render_video(element, context) + '\n'
            elif typ == 'map':
                html += cls.render_map(element, context) + '\n'
        return html

    @classmethod
    def _polygon_to_points(cls, polygon):
        polygon = polygon.replace('%', '')
        pairs = polygon.split(',')
        for index in range(len(pairs)):
            pair = pairs[index].strip()
            items = re.split('\s+', pair)
            for pos in range(len(items)):
                items[pos] = Decimal(items[pos]) / 100
                items[pos] = str(items[pos])
            pairs[index] = ' '.join(items)
        points = ', '.join(pairs)
        return points

    @classmethod
    def render_region(cls, poster_region, region):
        poster_region.class_name = 'p%s-p%s-%s' % (
            poster_region.poster_page.poster.id,
            poster_region.poster_page.index,
            poster_region.name
        )
        poster_region.element_id = poster_region.class_name
        poster_region.path_id = poster_region.element_id + '-path'
        poster_region.points = cls._polygon_to_points(poster_region.polygon)
        widget = region.get('widget')
        if widget:
            cls.render_widget(poster_region, widget)

    @classmethod
    def render_widget(cls, poster_region, widget):
        widget['class_name'] = '%s-%s' % ('poster-widget', widget['name'])
        widget['element_id'] = widget['class_name']
        typ = widget['type']
        context = dict(object=poster_region.poster_page.poster)
        if typ == 'background':
            html = cls.render_background(widget, context)
        # elif typ == 'text':
        #    html = cls.render_text(widget, context)
        elif typ == 'button':
            html = cls.render_button(widget, context)
        elif typ == 'image':
            html = cls.render_image(widget, context)
        elif typ == 'slider':
            html = cls.render_slider(widget, context)
        elif typ == 'music':
            html = cls.render_music(widget, context)
        elif typ == 'video':
            html = cls.render_video(widget, context)
        elif typ == 'map':
            html = cls.render_map(widget, context)
        else:
            html = ''
        widget['content'] = html
        poster_region.widget = widget

    @classmethod
    def render_text_widget(cls, poster, widget):
        widget['class_name'] = '%s-%s' % ('poster-widget', widget['name'])
        widget['element_id'] = widget['class_name']
        context = dict(object=poster)
        widget['content'] = cls.render_text(widget, context)

    @classmethod
    def render_background(cls, widget, context):
        template_name = 'libs/widgets/default.html'
        images = context['object'].images
        image_name = widget['image_name']
        if image_name in images:
            widget['image_url'] = images[image_name].file.url
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_text(cls, widget, context):
        template_name = 'libs/widgets/text.html'
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_button(cls, widget, context):
        template_name = 'libs/widgets/button.html'
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_image(cls, widget, context):
        template_name = 'libs/widgets/image.html'
        images = context['object'].images
        image_name = widget['image_name']
        if image_name in images:
            widget['image_url'] = images[image_name].file.url
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_slider(cls, widget, context):
        template_name = 'libs/widgets/slider.html'
        object_images = context['object'].images
        element_images = widget['images']
        for element_image in element_images:
            image_name = element_image['image_name']
            if image_name in object_images:
                image = object_images[image_name]
                element_image['image_url'] = image.file.url
                element_image['width'] = image.width
                element_image['height'] = image.height
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_music(cls, widget, context):
        template_name = 'libs/widgets/music.html'
        music = context['object'].music
        if music is not None:
            widget['music_url'] = music.file.url
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_video(cls, widget, context):
        template_name = 'libs/widgets/video.html'
        videos = context['object'].videos
        video_name = widget['video_name']
        widget['video_element_id'] = 'video-%s' % video_name
        if video_name in videos:
            widget['video'] = video = videos[video_name]
            video.url = video.file.url
        context['widget'] = widget
        return render_to_string(template_name, context)

    @classmethod
    def render_map(cls, widget, context):
        template_name = 'libs/widgets/map.html'
        context['widget'] = widget
        return render_to_string(template_name, context)

File: alatting_website/static/website/js/poster_render.js
/**
 * Created by ubuntu on 11/3/15.
 */
var PosterRender = {}

PosterRender.process = function(poster){
    this.init(poster)
    this.render(poster, poster.root)
}

PosterRender.init = function(object){
    //images
    var images = {}
    for(var index in object.poster_images){
        var poster_image = object.poster_images[index]
        images[poster_image.name] = poster_image.image
    }
    object.images = images
    //videos
    var videos = {}
    for(var index in object.poster_videos){
        var poster_video = object.poster_videos[index]
        videos[poster_video.name] = poster_video.video
    }
    object.videos = videos
    //pages, regions, texts
    var pages = []
    var regions = []
    for(var index in object.poster_pages){
        var poster_page = object.poster_pages[index]
        poster_page.poster = object
        pages[poster_page.index] = poster_page
        var page_regions = []
        var template_regions = poster_page.template.template_regions
        for(var position in template_regions) {
            var template_region = template_regions[position]
            template_region.poster_page = poster_page
            page_regions.push(template_region)
            regions.push(template_region)
        }
        poster_page.regions = page_regions
        //text
        for(var position in poster_page.page_texts) {
            poster_page.page_texts[position] = poster_page
        }
        poster_page.texts = poster_page.page_texts
    }
    object.pages = pages
    object.regions = regions
    return object
}

PosterRender.render = function(poster, root){
    for(var i = 0; i < poster.pages.length && i < root.pages.length; ++i){
        var poster_page = poster.pages[i]
        var page = root.pages[i]
        for(var j = 0; j < poster_page.regions.length; ++j){
            var page_region = poster_page.regions[j]
            var region = page.regions[page_region.name]
            if(region){
                this.renderRegion(page_region, region)
            }
        }
        for(var j = 0; j < poster_page.texts.length; ++j){
            var page_text = poster_page.texts[j]
            var text = page.texts[page_text.name]
            if(text){
                this.renderTextWidget(page_text, text)
            }
        }
    }
}

PosterRender.renderRegion = function(page_region, region){
    page_region.class_name = 'p' + page_region.poster_page.poster.id + '-p' + page_region.poster_page.index + '-' +
        page_region.name
    page_region.element_id = page_region.class_name
    page_region.path_id = page_region.element_id + '-path'
    $.extend(page_region, region)
    var widget = region.widget
    if(widget) {
        this.renderWidget(page_region, widget)
    }
}

PosterRender.renderWidget = function(page_region, widget){
    widget['class_name'] = 'poster-widget-' + widget['name']
    widget['element_id'] = widget['class_name']
    var type = widget['type']
    var context = {'object': page_region.poster_page.poster}
    if(type == 'image')
        this.renderImage(widget, context)
    else if(type == 'slider')
        this.renderSlider(widget, context)
    else if(type == 'music')
        this.renderMusic(widget, context)
    else if(type == 'video')
        this.renderVideo(widget, context)
}

PosterRender.renderTextWidget = function(page_text, text) {
    text['class_name'] = 'poster-widget-' + text['name']
    text['element_id'] = text['class_name']
    var context = {}
    $.extend(page_text, text)
    this.renderText(page_text, context)
}

PosterRender.renderText = function(widget, context) {
}

PosterRender.renderImage = function(widget, context) {
    var images = context['object'].images
    var image_name = widget['image_name']
    if(image_name in images) {
        widget['image_url'] = images[image_name].file
    }
    widget.include = imageWidgetURL
}

PosterRender.renderSlider = function(widget, context) {
    var object_images = context['object'].images
    var element_images = widget['images']
    for(var i=0; i < element_images.length; ++i){
        var element_image = element_images[i]
        var image_name = element_image['image_name']
        if(image_name in object_images) {
            var image = object_images[image_name]
            element_image['image_url'] = image.file
            element_image['width'] = image.width
            element_image['height'] = image.height
        }
    }
    widget.include = sliderWidgetURL
    widget.init = function(){
        initFlexSlider('#' + widget.element_id)
    }
}

PosterRender.renderMusic = function(widget, context) {
    var music = context['object'].music
    if(music) {
        widget['music_url'] = music.file
    }
    widget.include = musicWidgetURL
}

PosterRender.renderVideo = function(widget, context) {
    var videos = context['object'].videos
    var video_name = widget['video_name']
    widget['video_element_id'] = 'video-' + video_name
    if(video_name in videos) {
        widget['video'] = videos[video_name]
        widget['video'].url = widget['video'].file
    }
    widget.include = videoWidgetURL
    widget.init = function(){
        videojs.initialize(widget.video_element_id)
    }
}
